File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let Some(root) = root else {
            return 0;
        };

        let mut stack = vec![(root, 1)];
        let mut answer = 0;
        while let Some((node, depth)) = stack.pop() {
            answer = answer.max(depth);
            let borrowed = node.borrow();
            if let Some(left) = borrowed.left.clone() {
                stack.push((left, depth + 1));
            }
            if let Some(right) = borrowed.right.clone() {
                stack.push((right, depth + 1));
            }
        }
        answer
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
class Solution:
    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        stack = []
        node = root

        while True:
            while node is not None:
                stack.append(node)
                node = node.left

            node = stack.pop()
            k -= 1
            if k == 0:
                return node.val
            node = node.right